feat(ios, android): FLUID size support

Gregoirevda opened the pull request:
### Description

Support for FLUID size on Android and fix for iOS FLUID size.

Android: FLUID size wasn't supported yet on Android. Fixes #120 
iOS: FLUID size was taking full device width. Fixes #213 

### Checklist

- I read the [Contributor Guide](https://github.com/invertase/react-native-google-mobile-ads/blob/main/CONTRIBUTING.md)
  and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `__tests__e2e__`
  - [ ] `jest` tests added or updated in `__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---
Tested on various sizes including FLUID on Android and iOS, with different media (video/poll/creatives).

Think `react-native-google-mobile-ads` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`Invertase`](https://twitter.com/invertaseio) on Twitter

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-google-mobile-ads?pullRequest=277) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-google-mobile-ads?pullRequest=277) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-google-mobile-ads?pullRequest=277) it.</sub>

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:
```diff
@@ -123,7 +126,7 @@ public void setSizes(ReactViewGroup reactViewGroup, ReadableArray value) {
     }
 
-    if (sizeList.size() > 0) {
+    if (sizeList.size() > 0 && !sizeList.contains(AdSize.FLUID)) {
```
No need to send dimensions to RN as RN View is set as `flex: 1` and the native view can now layout itself

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:
```diff
@@ -0,0 +1,59 @@
+import com.google.android.gms.ads.BaseAdView;
+
+public class LayoutWrapper extends FrameLayout {
+  private BaseAdView contentView;
```
the `ViewGroup` can only contain a single `BaseAdView`

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:
```diff
@@ -0,0 +1,59 @@
+  @Override
+  public void requestLayout() {
+    super.requestLayout();
+    post(measureAndLayout);
```
We're sending our own `measure` and `layout`

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:
```diff
@@ -0,0 +1,59 @@
+
+    BaseAdView adView = getContentView();
+
+    if(adView != null) {
```
If there's an ad set, measure and layout based on ad dimensions

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:
```diff
@@ -0,0 +1,59 @@
+        MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY));
+      layout(left, top, left + width, top + height);
+    } else {
+      measure(
```
If no ad set (requestLayout called before it was initialized), layout and measure based on ViewGroup

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:
```diff
@@ -0,0 +1,59 @@
+    addView(contentView);
+  }
+
+  public BaseAdView getContentView() {
```
Remove the need of private method in `ReactNativeGoogleMobileAdsBannerAdViewManager.java`

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:
```diff
@@ -38,14 +39,16 @@
 
 public class ReactNativeGoogleMobileAdsBannerAdViewManager
-    extends SimpleViewManager<ReactViewGroup> {
+  extends SimpleViewManager<LayoutWrapper> {
```
ReactViewGroup overrides the [requestLayout](https://github.com/facebook/react-native/blob/0.66-stable/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java#L169), so extending `FrameLayout` which can only have a single child `BaseAdView` and setting our `requestLayout`

Gregoirevda commented on android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:
```diff
@@ -164,85 +167,68 @@ private BaseAdView initAdView(ReactViewGroup reactViewGroup) {
+      new AdListener() {
+        @Override
+        public void onAdLoaded() {
+          reactViewGroup.requestLayout();
```
requestLayout after ad loaded, no need to send dimensions when size is FLUID

Gregoirevda commented:
You actually need to wrap your `<GAMBannerAd` in a view with `height` and `width` eg
```jsx
<View style={{minHeight: 100, width: '100%'}}>
 <GAMBannerAd unitId="" sizes="FLUID" />
</View>
```

mikehardy commented:
Hey! :wave: thanks for posting this - I was out traveling but I'll be back at it next week / during work week and I'll check through this. I appreciate the patience :pray:

Gregoirevda commented:
Hi @mikehardy thank you for the update. Enjoy your holidays

mikehardy commented on src/ads/BaseAd.tsx:
```diff
@@ -115,8 +115,7 @@ export const BaseAd = React.forwardRef<GoogleMobileAdsBannerView, GAMBannerAdPro
       ? {
-          width: '100%',
-          height: dimensions[1],
+          flex: 1
```
@caiqueex this PR may be interesting for you if you want to take a look and help review

caiqueex replied:
Hi Mike, I o the test with this PR code. the result was it:

<img width="361" alt="Captura de Tela 2022-12-10 às 23 26 20" src="https://user-images.githubusercontent.com/8368079/206883576-e7f37450-210a-4261-9abb-723d1cedf30a.png">

I dont know why,  but it has that black background color and content to the right side

caiqueex replied:
<img width="277" alt="Captura de Tela 2022-12-10 às 23 31 22" src="https://user-images.githubusercontent.com/8368079/206883668-9ba28f0d-9ff7-4906-9833-43dd8342d647.png">

@mikehardy

Gregoirevda replied:
TestId.BANNER has 320x50. Does the black area represent 320+? Could you make a native ad on GAM and give it a 100% width and fixed height?

Gregoirevda commented:
@mikehardy Anything I can do to help this being shipped?

mikehardy commented:
Hey - sorry for the delay - the code seems okay but I'd love to see a snippet added to the example app that exercises this - I don't see where we are using the fluid style anywhere in the example app and there's no way for me to see it's working without an addition there

dylancom commented:
Do we actually need "FLUID" support? The Mobile Ads SDK docs seems to solely focus on using the new anchored & inline adaptive banner ads. The problem of our current implementation is that an inline adaptive banner takes the screen's width instead of for e.g. the container views width.
https://developers.google.com/admob/ios/banner

dylancom commented:
I made a PR that supports ads with custom width + auto height (using the new inline adaptive banner format). Would this make the FLUID size obsolete? https://github.com/invertase/react-native-google-mobile-ads/pull/296

Gregoirevda commented:
This is to support native styles ads. They can only be fixed or fluid.

https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size

dylancom commented:
Seems currently confusing for people like @caiqueex that actually need an adaptive banner with a custom width.

MateusAndrade commented:
@Gregoirevda did you manage to render FLUID ads on Android/iOS with this patch?

Gregoirevda commented:
@MateusAndrade yes

MateusAndrade commented:
I see, I will test this patch then. I believe this is something worth it having in a new release. Should we open the pr again?

mikehardy commented:
Hey @MateusAndrade 👋 😄 - happy to reopen this if it's useful for people. It's not a use case of mine so I'll need to know it is working, but we can merge it if it is. I fear there may be some conflicts as it has been closed a while - those will need resolution as well

MateusAndrade commented:
> Hey @MateusAndrade 👋 😄 - happy to reopen this if it's useful for people. It's not a use case of mine so I'll need to know it is working, but we can merge it if it is. I fear there may be some conflicts as it has been closed a while - those will need resolution as well

Nice, I see that most of the changes introduced with this PR were using Java/Objective-C so I believe we will need to change this to Kotlin/Swift.

Btw, I was testing the iOS implementation but I didn't manage to make it. The AD was not showing, any clue @Gregoirevda ?

Gregoirevda commented:
I've changed project meanwhile. Did you set a fixed width/height in the component? (windowWidth and fixed height for example) Can you share a code snippet

MateusAndrade commented:
> I've changed project meanwhile. Did you set a fixed width/height in the component? (windowWidth and fixed height for example) Can you share a code snippet

Not really, I was testing using the sizes provided by library and indeed they work. But when I use `FLUID` on Android it doesn't render at all, while on iOS it renders but the AD doesn't fit in the slot, it ends overflowing it's parent. 

```jsx
    const sizes =
        Platform.OS === 'ios'
            ? ["FLUID"]
            : [BannerAdSize.MEDIUM_RECTANGLE, '336x280'];

    return (
                        <View>
                            <GAMBannerAd
                                unitId={AD_UNIT_ID}
                                sizes={sizes}
                                requestOptions={{
                                    customTargeting,
                                    publisherProvidedId,
                                }}
                            />
                        </View>
)
```

Where my `AD_UNIT` is either a real ad id or a test id. Should I use fixed dimensions for the AD container?

Gregoirevda commented:
@MateusAndrade Not sure if it was on the `GAMBannerAd` or container `View`, but it was required.

[this](https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size) is what you currently have I think

MateusAndrade commented:
> @MateusAndrade Not sure if it was on the `GAMBannerAd` or container `View`, but it was required.
> 
> [this](https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size) is what you currently have I think

I see, then we definitely got a bug here. Even if the parent component has fixed dimensions, the scaling to fit is not working. 😭

Rajesh2github commented:
Hi when the FLUID ads size will work when we are planning to merge this PR

mikehardy commented:
@Rajesh2github "we" ? Good question - I don't recall seeing you contribute to the PR. It looks like it has issues and if you need it you could work on the PR to fix the issues. We'd be happy to merge it at that point

mikehardy commented:
:tada: This issue has been resolved in version 13.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/react-native-google-mobile-ads/v/13.2.0)
- [GitHub release](https://github.com/invertase/react-native-google-mobile-ads/releases/tag/v13.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
